libs/dal/src/repositories/base-repository.ts

```diff
@@ -1,15 +1,28 @@
-/* eslint-disable @typescript-eslint/no-explicit-any */
 import { ClassConstructor, plainToInstance } from 'class-transformer';
 import { addDays } from 'date-fns';
 import {
   MESSAGE_GENERIC_RETENTION_DAYS,
   MESSAGE_IN_APP_RETENTION_DAYS,
   NOTIFICATION_RETENTION_DAYS,
 } from '@novu/shared';
-import { Model, Types, ProjectionType, FilterQuery, UpdateQuery, QueryOptions } from 'mongoose';
-import { DalException } from '../shared';
-
-export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
+import { Model, Types, ProjectionType, QueryOptions, FilterQuery, UpdateQuery } from 'mongoose';
+import { ObjectIdType } from '../types';
+
+/**
+ * The query type used for enforcing the type of the query data payload
+ */
+type QueryData<T_DBModel, T_Enforcement> = FilterQuery<T_DBModel> & T_Enforcement;
+
+/**
+ * The create type used for enforcing the type of the create data payload
+ */
+type CreateData<T_DBModel, T_Enforcement> = Omit<T_DBModel, '_id'> & T_Enforcement;
```

Comment:
I think we should add here some more params to Omit like all the date ones, `createdAt, updatedAt` maybe the `deleted` as well. There could be a couple of more as well.

Reply:
Good catch. I agree that some date attributes should be removed, however there are cases where we might like to customise the property outside the DAL. A good example here is the `expireAt` property,  which we often specify as part of an `insert` operation.

We could find a middle ground by selecting all date attributes via the Date template literal type, then make them optional via Partial.

@LetItRock what do you think?

Reply:
I have added `IEntity` and `IDBModel` as opposed to a base class, to give us more flexibility when decorating the `_id` attribute. This solution also enables us to use custom `_id` string types when implementing the Entity class.

Let me know what you think @djabarovgeorge